[MAGETWO-54158]:+ Corrected ACL rules for the sendfriend module and s…

jspoe opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->
Added an acl.xml to the send-friend module
Added upgrade data script for backward compatibility for ACL-roles that were already selected for users.

Changed the acl.xml for the backend module
Added upgrade data script for backward compatibility for ACL-roles that were already selected for users.

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. magento/magento2#MAGETWO-54158: ACL not defined for Magento_Backend::web

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. Install Magento latest develop version
2. Go to Admin side
3. Click "System > User Roles"
4. Click "Add new Role" enter Role name
5. In left side choose Role Resources and check "Dashboard", "Web Section", "Email to a friend", "Currency Setup", "Store E-mail addresses"
6. Save Role
7. Click "System > All Users"
8. Click "Add new User"
9. Enter data in required fields
10. In left side click "User Role" and choose early created role.
11. Click "Save"
12. Sign out and Log in using credentials of new user
13. Click "Stores > Configuration"

Expected result:
All the tabs that were selected in the role should be visible

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

jspoe committed:
[MAGETWO-54158]:+ Corrected ACL rules for the sendfriend module and several backend store configuration

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=9628) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=9628) before we can accept your contribution.<br/><hr/>**Johan Spoelstra** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

jspoe committed:
[MAGETWO-54158]:+ Travis codestyle

ishakhsuvarov commented on app/code/Magento/Config/Setup/UpgradeData.php:
```diff
@@ -0,0 +1,55 @@
+     * @param ModuleDataSetupInterface $setup
+     * @return void
+     */
+    protected function upgradeAcl(ModuleDataSetupInterface $setup)
```
Is there any reason for this method is protected?  Looks like private is more suitable in this case.

ishakhsuvarov commented on app/code/Magento/Config/Setup/UpgradeData.php:
```diff
@@ -0,0 +1,55 @@
+class UpgradeData implements UpgradeDataInterface
+{
+
+    /**
```
`{@inheritdoc}` is sufficient in this case.

ishakhsuvarov commented on app/code/Magento/Config/Setup/UpgradeData.php:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    /**
+     * @param ModuleDataSetupInterface $setup
```
Would be nice to add a comment which explains background under this action.

ishakhsuvarov commented on app/code/Magento/Config/etc/acl.xml:
```diff
@@ -12,16 +12,15 @@
                     <resource id="Magento_Backend::stores_settings">
                         <resource id="Magento_Config::config" title="Configuration" translate="title" sortOrder="20">
-                            <resource id="Magento_Config::advanced" title="Advanced Section" translate="title" sortOrder="90" />
+                            <resource id="Magento_Backend::advanced" title="Advanced Section" translate="title" sortOrder="90" />
```
Did you consider introducing these resources in the `Backend/etc/acl.xml`?

ishakhsuvarov commented on app/code/Magento/SendFriend/Setup/UpgradeData.php:
```diff
@@ -0,0 +1,45 @@
+     * @param ModuleDataSetupInterface $setup
+     * @return void
+     */
+    protected function upgradeAcl(ModuleDataSetupInterface $setup)
```
Please consider making this method private

ishakhsuvarov commented:
Hi @jspoe 
Would you like to continue work on this PR or should someone else pick it up?

jspoe commented:
Hi @ishakhsuvarov 

Thank you for your feedback!
I will pickup the requested changes in a couple of days. Should not take a lot of time.

jspoe committed:
[MAGETWO-54158]:+ Requested QR changes

+ Made setup functions private
+ Moved Backend ACL nodes to Backend module instead of Config module

ishakhsuvarov commented:
@jspoe Thanks, we will proceed with review shortly.

ishakhsuvarov commented:
@jspoe It looks like these changes introduce some unwanted dependencies between the modules in question. We will look into possible solution and let you know how we can fix it.

jspoe commented:
Yes I understand. It is either part of the Magento_Config module or the Magento_Backend module.
It should not be both. Then there is also the part of the setup script where there is a (small) dependency for to map the roles for backward compatibility.

Please let me know as soon as possible if I can fix this in a better way for you.

ishakhsuvarov commented:
@jspoe Looks like the best solution in this case would be to move upgrade scripts to the specifically created modules, something like SendfriendAcl or SendfriendAuthorization, which would have correct dependencies and would allow other modules to be disabled/removed safely.
Would you like to continue on that, since it is some more work to do, or do you want someone else to pick up?
Thanks!

jspoe commented:
Hi @ishakhsuvarov,

I agree, and I will pick up the requests myself if possible.
I will push a change as soon as I can.

Thanks!

ishakhsuvarov commented:
@jspoe Thank you

ishakhsuvarov commented:
@jspoe I am closing this PR for now due to inactivity. Please feel free to update and reopen whenever you wish to continue work on it.
Thank you!
